drivers/modem/modem_cellular.c

```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
 static void modem_cellular_await_power_on_event_handler(struct modem_cellular_data *data,
 							enum modem_cellular_event evt)
 {
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		if (config->set_baudrate_chat_script != NULL) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_SET_BAUDRATE);
+		} else {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_RUN_INIT_SCRIPT);
+		}
+		break;
+
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_set_baudrate_state_enter(struct modem_cellular_data *data)
+{
+	modem_pipe_attach(data->uart_pipe, modem_cellular_bus_pipe_handler, data);
+	return modem_pipe_open_async(data->uart_pipe);
+}
+
+static void modem_cellular_set_baudrate_event_handler(struct modem_cellular_data *data,
+						      enum modem_cellular_event evt)
+{
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+	struct uart_config cfg = {0};
+
 	switch (evt) {
+	case MODEM_CELLULAR_EVENT_BUS_OPENED:
+		modem_chat_attach(&data->chat, data->uart_pipe);
+		modem_chat_run_script_async(&data->chat, config->set_baudrate_chat_script);
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		/* Let modem reconfigure */
+		modem_cellular_start_timer(data, K_MSEC(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE_DELAY));
+		break;
+	case MODEM_CELLULAR_EVENT_SCRIPT_FAILED:
+		/* Some modems save the new speed on first change, meaning the
+		 * modem is already at the new baudrate, meaning no reply. So
+		 * ignore any failures and continue as if baudrate is already set
+		 */
+		__fallthrough;
 	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		uart_config_get(config->uart, &cfg);
+		cfg.baudrate = CONFIG_MODEM_CELLULAR_NEW_BAUDRATE;
+		uart_configure(config->uart, &cfg);
+
+		modem_chat_release(&data->chat);
+		modem_pipe_attach(data->uart_pipe, modem_cellular_bus_pipe_handler, data);
+		modem_pipe_close_async(data->uart_pipe);
```

Comment:
Would it make more sense to do the releasing/closing before changing the UART configuration? (not sure)

Reply:
Seems to work, shouldn't really make any practical difference

Reply:
Curious to know @bjarki-andreasen's viewpoint on this?